packages/mds-compliance-api/handlers/get-violation-periods.ts

```diff
@@ -0,0 +1,144 @@
+}>
+
+interface ComplianceAggregateMap {
+  [k: string]: {
+    complianceSnapshotViolationGroupings: ComplianceSnapshotDomainModel[][]
+    complianceSnapshots: null | ComplianceSnapshotDomainModel[]
+  }
+}
+
+async function convertComplianceSnapshotsArrayToComplianceViolationPeriod(
+  snapshots: ComplianceSnapshotDomainModel[]
+): Promise<ComplianceViolationPeriod> {
+  const compliance_array_response_id = uuid()
+  await ComplianceServiceClient.createComplianceArrayResponse({
+    compliance_array_response_id,
+    compliance_snapshot_ids: snapshots.map(s => s.compliance_snapshot_id),
+    provider_id: snapshots[0].provider_id
+  })
+  return {
+    start_time: snapshots[0].compliance_as_of,
+    end_time: snapshots[snapshots.length - 1].compliance_as_of,
+    snapshots_uri: `/compliance_snapshot_ids?token=${compliance_array_response_id}`
+  }
+}
+
+export const GetViolationPeriodsHandler = async (
+  req: ComplianceApiGetViolationPeriodsRequest,
+  res: ComplianceApiGetViolationPeriodsResponse,
+  next: express.NextFunction
+) => {
+  try {
+    const { scopes } = res.locals
+    const { start_time, end_time = now() } = parseRequest(req)
+      .single({ parser: Number })
+      .query('start_time', 'end_time')
+    if (!isDefined(start_time)) {
+      return res.status(400).send({ error: 'Missing required query param start_time' })
+    }
+    const { provider_ids, policy_ids } = parseRequest(req)
+      .single({ parser: (str: string | undefined | null) => str?.split(',') })
+      .query('provider_ids', 'policy_ids')
+
+    let p_provider_ids
+    if (scopes.includes('compliance:read:provider')) {
+      if (res.locals.claims && res.locals.claims.provider_id) {
+        const { provider_id } = res.locals.claims
+        p_provider_ids = [provider_id]
+      } else {
+        return res.status(403).send({ error: 'compliance:read:provider token missing provider_id' })
+      }
+    } else if (scopes.includes('compliance:read')) {
+      p_provider_ids = provider_ids ?? Object.keys(providers)
+    }
+
+    const p_policy_ids = policy_ids ?? (await db.readActivePolicies()).map((p: Policy) => p.policy_id)
+    const complianceSnapshots = await ComplianceServiceClient.getComplianceSnapshotsByTimeInterval({
+      start_time,
+      end_time,
+      provider_ids: p_provider_ids,
+      policy_ids: p_policy_ids
+    })
+
+    const complianceAggregateMap: ComplianceAggregateMap = {}
+
+    /**
+     * Iterate through all compliance snapshots. For each policy-provider pair, build up arrays of
+     * contiguous compliance snapshots that contain violations. End the array when a compliance snapshot
+     * is encountered that has no violations. E.g. if for Jump and policy A, there are snapshots B, C, D, E,
+     * and F, and B, C, E and F contain violations, B and C get grouped together and eventually put into the
+     * same instance of a ComplianceViolationPeriod, and E and F get grouped together. D is basically ignored.
+     *
+     */
```

Comment:
This *definitely* feels like service-level logic, and _perhaps_ even repository-level (there's probably a way more performant way to do this in SQL)

Reply:
Yeah, fair given how thin the API should be. I'll move it over.

Reply:
Currently waiting on Kym to investigate the plausibility of doing this all in SQL.